source/normalize-arguments.ts

```diff
@@ -103,143 +120,290 @@ export const preNormalizeArguments = (options: Options, defaults?: NormalizedOpt
+		}
+	} else if (isForm) {
+		if (!is.object(options.form)) {
+			throw new TypeError('The `form` option must be an Object');
+		}
+
+		if (!Reflect.has(headers, 'content-type')) {
+			headers['content-type'] = 'application/x-www-form-urlencoded';
+		}
+
+		options.body = (new URLSearchParams(options.form as Record<string, string>)).toString();
+	} else if (isJSON) {
+		if (!Reflect.has(headers, 'content-type')) {
+			headers['content-type'] = 'application/json';
+		}
+
+		options.body = JSON.stringify(options.json);
+	}
+
+	// Convert buffer to stream to receive upload progress events (#322)
+	if (is.buffer(options.body)) {
+		uploadBodySize = options.body.length;
+		options.body = toReadableStream(options.body);
+	} else {
+		uploadBodySize = await getBodySize(options);
+	}
 
-			searchParams = (new URLSearchParams(searchParams as Record<string, string>)).toString();
+	// See https://tools.ietf.org/html/rfc7230#section-3.3.2
+	// A user agent SHOULD send a Content-Length in a request message when
+	// no Transfer-Encoding is sent and the request method defines a meaning
+	// for an enclosed payload body.  For example, a Content-Length header
+	// field is normally sent in a POST request even when the value is 0
+	// (indicating an empty payload body).  A user agent SHOULD NOT send a
+	// Content-Length header field when the request message does not contain
+	// a payload body and the method semantics do not anticipate such a
+	// body.
+	if (!Reflect.has(headers, 'content-length') && !Reflect.has(headers, 'transfer-encoding')) {
+		if (
+			(options.method === 'POST' || options.method === 'PUT' || options.method === 'PATCH') &&
+			!is.undefined(uploadBodySize)
+		) {
+			headers['content-length'] = String(uploadBodySize);
 		}
+	}
+
+	if (!options.isStream && options.responseType === 'json' && is.undefined(headers.accept)) {
+		headers.accept = 'application/json';
+	}
+
+	if (options.decompress && is.undefined(headers['accept-encoding'])) {
+		headers['accept-encoding'] = supportsBrotli ? 'gzip, deflate, br' : 'gzip, deflate';
+	}
+
+	// Validate URL
+	if (options.url.protocol !== 'http:' && options.url.protocol !== 'https:') {
+		throw new UnsupportedProtocolError(options);
+	}
 
-		options.path = `${options.path.split('?')[0]}?${searchParams}`;
+	decodeURI(options.url.toString());
+
+	// Normalize request function
+	if (!is.function_(options.request)) {
+		options.request = options.url.protocol === 'https:' ? https.request : http.request;
 	}
 
-	if (options.hostname === 'unix') {
-		const matches = /(?<socketPath>.+?):(?<path>.+)/.exec(options.path);
+	// UNIX sockets
+	if (options.url.hostname === 'unix') {
+		const matches = /(?<socketPath>.+?):(?<path>.+)/.exec(options.url.pathname);
 
 		if (matches?.groups) {
 			const {socketPath, path} = matches.groups;
+
+			// It's a bug!
+			// eslint-disable-next-line @typescript-eslint/consistent-type-assertions
```

Comment:
@sindresorhus can you take a look at this?

Reply:
Yeah, looks like a bug in typescript-eslint. It's so buggy I don't want to spend time even opening an issue. Let's leave it like this and open an issue there sometime in the future if it's still not fixed.